src/cstring.rs

```diff
@@ -0,0 +1,450 @@
+    #[inline]
+    unsafe fn pop_terminator(&mut self) {
+        if cfg!(debug_assertions) {
+            let popped = self.buf.pop();
+
+            assert_eq!(popped, Some(0));
+        } else {
+            unsafe { self.buf.pop_unchecked() };
+        }
+    }
+
+    /// Converts this [`CString`] to a string slice if the [`CString`] represents valid UTF-8.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use heapless::CString;
+    ///
+    /// let mut cstr: CString<10> = CString::new();
+    /// cstr.push_bytes(b"heapless").unwrap();
+    ///
+    /// assert_eq!(
+    ///     cstr.to_str(),
+    ///     Ok("heapless")
+    /// );
+    /// ```
+    pub fn to_str(&self) -> Result<&str, core::str::Utf8Error> {
+        core::str::from_utf8(self.inner_without_nil_terminator())
+    }
+
+    /// Converts a [`CString`] to a string slice without checking
+    /// that the string contains valid UTF-8.
+    ///
+    /// See the safe version, [`CString::to_str`], for more information.
+    ///
+    /// # Safety
+    ///
+    /// The bytes passed in must be valid UTF-8.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use heapless::CString;
+    ///
+    /// let mut cstr: CString<10> = CString::new();
+    /// cstr.push_bytes(b"heapless").unwrap();
+    ///
+    /// assert_eq!(
+    ///     unsafe { cstr.as_str_unchecked() },
+    ///     "heapless",
+    /// );
+    /// ```
+    pub unsafe fn as_str_unchecked(&self) -> &str {
+        core::str::from_utf8_unchecked(self.inner_without_nil_terminator())
+    }
+
+    /// Removes the existing nul terminator and then extends `self` with the given bytes.
+    ///
+    /// # Safety
+    ///
+    ///  If `additional` is not null-terminated, the CString is left non null-terminated, which is
+    ///  an invalid state. Caller must ensure that either `additional` has a terminating nul byte
+    ///  or ensure to fix the CString after calling `extend_slice`.
+    unsafe fn extend_slice(&mut self, additional: &[u8]) -> Result<(), ()> {
+        self.pop_terminator();
+
+        self.buf.extend_from_slice(additional)
+    }
+
+    #[inline]
+    fn inner_without_nil_terminator(&self) -> &[u8] {
+        // Assert our invariant: `self.buf` must be null-terminated
+        debug_assert!(self.buf.len() > 0);
+        debug_assert_eq!(self.buf.last().copied(), Some(0));
+
+        &self.buf[..self.buf.len() - 1]
+    }
+
+    /// Gets the underlying bytes written to this [`CString`] so far, without including the null terminator.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use heapless::CString;
+    ///
+    /// let mut cstr = CString::<5>::new();
+    /// cstr.push_bytes(b"ab").unwrap();
+    /// cstr.push_bytes(b"cd").unwrap();
+    ///
+    /// assert_eq!(
+    ///     cstr.as_bytes_with_null(),
+    ///     b"abcd\0"
+    /// );
+    /// ```
+    #[inline]
+    pub fn as_bytes_with_null(&self) -> &[u8] {
```

Comment:
```suggestion
    pub fn as_bytes_with_nul(&self) -> &[u8] {
```

to be consistent with `std`.